// @f1api/sdk/dist/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  Circuits: () => Circuits,
  Drivers: () => Drivers,
  F1Api: () => F1Api,
  Races: () => Races,
  Results: () => Results,
  Seasons: () => Seasons,
  Standings: () => Standings,
  Teams: () => Teams
});
module.exports = __toCommonJS(index_exports);

// src/base.ts
var import_isomorphic_unfetch = __toESM(require("isomorphic-unfetch"));
var Base = class {
  constructor() {
    this.baseUrl = "https://f1api.dev/api";
  }
  async request(endpoint, options) {
    const url = `${this.baseUrl}${endpoint}`;
    const headers = {
      "Content-Type": "application/json"
    };
    const config = __spreadProps(__spreadValues({}, options), {
      headers
    });
    const response = await (0, import_isomorphic_unfetch.default)(url, config);
    if (response.ok) {
      return response.json();
    }
    throw new Error(response.statusText);
  }
};

// src/utils.ts
function applyMixins(derivedCtor, baseCtors) {
  baseCtors.forEach((baseCtor) => {
    Object.getOwnPropertyNames(baseCtor.prototype).forEach((name) => {
      const descriptor = Object.getOwnPropertyDescriptor(
        baseCtor.prototype,
        name
      );
      if (descriptor) {
        Object.defineProperty(derivedCtor.prototype, name, descriptor);
      }
    });
  });
}
function buildQueryParams(limit, offset) {
  const params = {};
  if (limit !== void 0) params.limit = String(limit);
  if (offset !== void 0) params.offset = String(offset);
  const queryString = new URLSearchParams(params).toString();
  return queryString;
}

// src/circuits/circuits.ts
var resourceName = "circuits";
var Circuits = class extends Base {
  getCircuits({
    limit,
    offset
  } = {}) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${resourceName}?${queryString}`;
    return this.request(url);
  }
  getCircuit({ id }) {
    const url = `/${resourceName}/${id}`;
    return this.request(url);
  }
  getCircuitsByYear({
    year,
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${year}/${resourceName}?${queryString}`;
    return this.request(url);
  }
};

// src/drivers/drivers.ts
var resourceName2 = "drivers";
var Drivers = class extends Base {
  // returns all the f1 drivers in the database
  getDrivers({
    limit,
    offset
  } = {}) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${resourceName2}?${queryString}`;
    return this.request(url);
  }
  // returns all the f1 drivers of a specific year
  getDriversByYear({
    year,
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${year}/${resourceName2}?${queryString}`;
    return this.request(url);
  }
  // returns a specific f1 driver based on his id
  getDriver({ id }) {
    const url = `/${resourceName2}/${id}`;
    return this.request(url);
  }
  // returns all the f1 drivers of the current year
  getCurrentDrivers({
    limit,
    offset
  } = {}) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/current/${resourceName2}?${queryString}`;
    return this.request(url);
  }
  // returns a specific f1 driver of the current year based on his id, with his results
  getCurrentDriver({
    id,
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/current/${resourceName2}/${id}?${queryString}`;
    return this.request(url);
  }
  // returns a f1 driver based on a year and his id, with his results
  getDriverByYear({
    year,
    id,
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${year}/${resourceName2}/${id}?${queryString}`;
    return this.request(url);
  }
};

// src/races/races.ts
var Races = class extends Base {
  // returns all the races of a specific year
  getRacesByYear({
    year,
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${year}?${queryString}`;
    return this.request(url);
  }
  // retruns all the races of the current year
  getCurrentRaces({
    limit,
    offset
  } = {}) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/current?${queryString}`;
    return this.request(url);
  }
  // returns details of a specific race based on the year and the round
  getRaceInfo({
    year,
    round
  }) {
    const url = `/${year}/${round}`;
    return this.request(url);
  }
  // returns info of the last race of the current season
  getLastRace() {
    const url = `/current/last`;
    return this.request(url);
  }
  // returns info of the next race of the current season
  getNextRace() {
    const url = `/current/next`;
    return this.request(url);
  }
};

// src/results/results.ts
var Results = class extends Base {
  //retrieves fp1 results of a race, based on the year and round
  getFp1Results({
    year,
    round,
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${year}/${round}/fp1?${queryString}`;
    return this.request(url);
  }
  //retrieves fp1 results of the last race of the current season
  getLastFp1Results({
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/current/last/fp1?${queryString}`;
    return this.request(url);
  }
  //retrieves fp2 results of a race, based on the year and round
  getFp2Results({
    year,
    round,
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${year}/${round}/fp2?${queryString}`;
    return this.request(url);
  }
  //retrieves fp2 results of the last race of the current season
  getLastFp2Results({
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/current/last/fp2?${queryString}`;
    return this.request(url);
  }
  //retrieves fp3 results of a race, based on the year and round
  getFp3Results({
    year,
    round,
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${year}/${round}/fp3?${queryString}`;
    return this.request(url);
  }
  //retrieves fp3 results of the last race of the current season
  getLastFp3Results({
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/current/last/fp3?${queryString}`;
    return this.request(url);
  }
  // retrieves qualifying results of a race, based on the year and round
  getQualyResults({
    year,
    round,
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${year}/${round}/qualy?${queryString}`;
    return this.request(url);
  }
  //retrieves qualifying results of the last race of the current season
  getLastQualyResults({
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/current/last/qualy?${queryString}`;
    return this.request(url);
  }
  //retrieves race results of a race, based on the year and round
  getRaceResults({
    year,
    round,
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${year}/${round}/race?${queryString}`;
    return this.request(url);
  }
  //retrieves race results of the last race of the current season
  getLastRaceResults({
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/current/last/race?${queryString}`;
    return this.request(url);
  }
  //retrieves sprint results of a race, based on the year and round
  getSprintRaceResults({
    year,
    round,
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${year}/${round}/sprint/race?${queryString}`;
    return this.request(url);
  }
  //retrieves sprint results of the last race of the current season
  getLastSprintRaceResults({
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/current/last/sprint/race?${queryString}`;
    return this.request(url);
  }
  //retrieves sprint qualy results of a race, based on the year and round
  getSprintQualyResults({
    year,
    round,
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${year}/${round}/sprint/qualy?${queryString}`;
    return this.request(url);
  }
  //retrieves sprint qualy results of the last race of the current season
  getLastSprintQualyResults({
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/current/last/sprint/qualy?${queryString}`;
    return this.request(url);
  }
};

// src/seasons/seasons.ts
var resourceName3 = "seasons";
var Seasons = class extends Base {
  getSeasons({
    limit,
    offset
  } = {}) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${resourceName3}?${queryString}`;
    return this.request(url);
  }
};

// src/standings/standings.ts
var driverStandings = "drivers-championship";
var constructorStandings = "constructors-championship";
var Standings = class extends Base {
  getDriverStandings({
    year,
    limit,
    offset
  }) {
    const queryParams = buildQueryParams(limit, offset);
    const url = `/${year}/${driverStandings}?${queryParams}`;
    return this.request(url);
  }
  getCurrentDriverStandings({
    limit,
    offset
  } = {}) {
    const queryParams = buildQueryParams(limit, offset);
    const url = `/current/${driverStandings}?${queryParams}`;
    return this.request(url);
  }
  getConstructorStandings({
    year,
    limit,
    offset
  }) {
    const queryParams = buildQueryParams(limit, offset);
    const url = `/${year}/${constructorStandings}?${queryParams}`;
    return this.request(url);
  }
  getCurrentConstructorStandings({
    limit,
    offset
  } = {}) {
    const queryParams = buildQueryParams(limit, offset);
    const url = `/current/${constructorStandings}?${queryParams}`;
    return this.request(url);
  }
};

// src/teams/teams.ts
var resourceName4 = "teams";
var Teams = class extends Base {
  // returns all the f1 teams in the database
  getTeams({
    limit,
    offset
  } = {}) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${resourceName4}?${queryString}`;
    return this.request(url);
  }
  //returns a f1 team based on his id
  getTeam({ id }) {
    const url = `/${resourceName4}/${id}`;
    return this.request(url);
  }
  //returns all the f1 teams of a specific year
  getTeamsByYear({
    year,
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${year}/${resourceName4}?${queryString}`;
    return this.request(url);
  }
  //returns all the f1 teams of the curren year
  getCurrentTeams({
    limit,
    offset
  } = {}) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/current/${resourceName4}?${queryString}`;
    return this.request(url);
  }
  // returns a team based on a year and his id
  getTeamByYear({
    year,
    id
  }) {
    const url = `/${year}/${resourceName4}/${id}`;
    return this.request(url);
  }
  // returns a team based on the current year and his id
  getCurrentTeam({ id }) {
    const url = `/current/${resourceName4}/${id}`;
    return this.request(url);
  }
  //returns a team with his drivers based on a year and the team id
  getTeamByYearWithDrivers({
    year,
    id,
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/${year}/${resourceName4}/${id}/drivers?${queryString}`;
    return this.request(url);
  }
  // returns a team with his drivers based on the current year and the team id
  getCurrentTeamWithDrivers({
    id,
    limit,
    offset
  }) {
    const queryString = buildQueryParams(limit, offset);
    const url = `/current/${resourceName4}/${id}/drivers?${queryString}`;
    return this.request(url);
  }
};

// src/index.ts
var F1Api = class extends Base {
};
applyMixins(F1Api, [
  Seasons,
  Circuits,
  Standings,
  Drivers,
  Teams,
  Races,
  Results
]);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Circuits,
  Drivers,
  F1Api,
  Races,
  Results,
  Seasons,
  Standings,
  Teams
});
